esprima.js

```diff
@@ -2293,74 +2306,74 @@
                     expect(')');
                     value = parsePropertyFunction(param, token);
                 }
-                return node.finishProperty('set', key, value, false, false);
+                return node.finishProperty('set', key[0], key[1], value, false, false);
             }
-            if (match(':')) {
-                lex();
-                value = parseAssignmentExpression();
-                return node.finishProperty('init', id, value, false, false);
+        }
+
+        if (match('(')) {
+            value = parsePropertyMethodFunction();
+            return node.finishProperty('init', key[0], key[1], value, true, false);
+        }
+
+        return key;
+    }
+
+    function isProto(key, computed) {
+        return computed === false && (key.type === Syntax.Identifier && key.name === '__proto__' ||
+            key.type === Syntax.Literal && key.value === '__proto__');
+    }
+
+    function parseObjectProperty(hasProto) {
+        var token = lookahead, node = new Node(), methodOrPropertyName = parseMethodOrPropertyName(node), value;
+
+        switch (methodOrPropertyName.type) {
+        case Syntax.Property:
+            if (isProto(methodOrPropertyName.key, methodOrPropertyName.computed)) {
+                if (hasProto.value) {
+                    tolerateError(Messages.DuplicateProtoProperty);
+                } else {
+                    hasProto.value = true;
+                }
             }
-            if (match('(')) {
-                value = parsePropertyMethodFunction();
-                return node.finishProperty('init', id, value, true, false);
+            // finished
+            return methodOrPropertyName;
+
+        default:
+            // init property or short hand property.
+            if (isProto(methodOrPropertyName[0], methodOrPropertyName[1])) {
```

Comment:
I find this is a bit confusing, consider the previous call (for `Syntax.Property`) is using `key` and `computed` properties as the parameters, not array indexes.